### PR TITLE
Add guest_email to checkout view

### PR DESCRIPTION
**Pull request by crgwbr**
Add guest_email field to checkout view. This is needed for anonymous checkouts.

### CODE REVIEW

**Comment by specialunderwear** [comment]
The tests are failing, please fix the unit tests that don't count on this new field.

**Comment by crgwbr** [comment]
OK, back to you @specialunderwear. I added some additional tests as for anonymous checkouts.

File: oscarapi/tests/testcheckout.py [hunk 1]
```diff
@@ -176,16 +179,56 @@ def test_checkout_header(self):
...
+                "line4": "Kroekingen",
+                "notes": "Niet STUK MAKEN OK!!!!",
+                "phone_number": "+31 26 370 4887",
+                "postcode": "7777KK",
+                "state": "Gerendrecht",
+                "title": "Mr"
+            }
+        }
+
+        with mock.patch('django.conf.settings.OSCAR_ALLOW_ANON_CHECKOUT', True, create=True):
```
Comment (specialunderwear) [inline oscarapi/tests/testcheckout.py]: Why not:

with self.settings(OSCAR_ALLOW_ANON_CHECKOUT=True):

File: setup.py [hunk 2]
```diff
@@ -36,7 +36,8 @@
     install_requires=[
         'setuptools',
         'django-oscar>=1.1',
-        'djangorestframework>=3.1.0'
+        'djangorestframework>=3.1.0',
+        'mock>=1.3.0',
```
Comment (specialunderwear) [inline setup.py]: Not needed? Or should be in test extras

File: oscarapi/tests/testcheckout.py (hunk 1 above)
Comment (maerteijn) [inline oscarapi/tests/testcheckout.py]: +1 on that. Depending on `mock>=1.3.0` just for this already build-in functionality is not really needed.

Comment (crgwbr) [inline oscarapi/tests/testcheckout.py]: Huh, I've never seen that before. Will change now

File: oscarapi/serializers/checkout.py [hunk 3]
```diff
@@ -182,6 +182,7 @@ class Meta:
 class CheckoutSerializer(serializers.Serializer, OrderPlacementMixin):
     basket = serializers.HyperlinkedRelatedField(
         view_name='basket-detail', queryset=Basket.objects)
+    guest_email = serializers.EmailField(allow_blank=True, required=False)
```
Comment (maerteijn) [inline oscarapi/serializers/checkout.py]: Why would you add `allow_blank=True`? Now it's possible to do an anonymous checkout without a valid email address

File: oscarapi/serializers/checkout.py [hunk 4]
```diff
@@ -195,11 +196,21 @@ def get_initial_order_status(self, basket):
...
+        if request.user.is_anonymous():
+            if 'guest_email' not in attrs:
+                # Always require the guest email field if the user is anonymous
+                message = _('Guest email is required for anonymous checkouts')
+                raise serializers.ValidationError(message)
+        else:
+            if 'guest_email' in attrs:
+                # Don't store guest_email field if the user is authenticated
+                attrs['guest_email'] = None
+
```
Comment (maerteijn) [inline oscarapi/serializers/checkout.py]: I would say to write it like this:
```python
        if request.user.is_anonymous():
            if not settings.OSCAR_ALLOW_ANON_CHECKOUT:
                message = _('Anonymous checkout forbidden')
                raise serializers.ValidationError(message)

            if 'guest_email' not in attrs:
                # always require the guest email field if the user is anonymous
                message = _('Guest email is required for anonymous checkouts')
                raise serializers.ValidationError(message)
        else:
            # delete guest_email when it's there because we are authenticated
            if 'guest_email' in attrs:
                del attrs['guest_email']
```

**Comment by crgwbr** [comment]
@maerteijn ok I addressed your concerns. I've added the allow blank param because some JS frameworks will send the param as an empty string and the only time that's actually an issue is when the user is anonymous. The validation logic now reflects this.

**Comment by maerteijn** [comment]
Looks a lot cleaner and clearer now. The only think that's missing in`def test_anonymous_checkout` is to test for an empty (or not specified) `guest_email` address. So I would say that you remove the `guest_email` from the request dictionary and write something like this:

``` python
with self.settings(OSCAR_ALLOW_ANON_CHECKOUT=True):
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 406)
    response = self.post('api-basket-add-product', url="http://testserver/api/products/1/", quantity=5)
    self.assertEqual(response.status_code, 200)

    # no guest email specified should say 406
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 406)

    # an empty email address should say this as well
    request['guest_email'] = ''
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 406)

    # a valid email address should say 200
    request['guest_email'] = 'henk@example.com'
    response = self.post('api-checkout', **request)
    self.assertEqual(response.status_code, 200)
    self.assertEqual(response.data['guest_email'], 'henk@example.com')
    self.assertEqual(Basket.objects.get(pk=basket_id).status, 'Frozen', 'Basket should be frozen after placing order and before payment')

**Comment by bufke** [comment]
@maerteijn - @crgwbr asked me to finish this up for him. I made the changes pretty as you suggested. Anything else needed?

**Comment by specialunderwear** [comment]
Is ok with me

**Comment by maerteijn** [comment]
Fine by me too.
